assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -3367,38 +3720,117 @@ public void doPrev_back_next_submission(RunData rundata, String option)
 		
 		if (state.getAttribute(STATE_MESSAGE) == null)
 		{
-			if ("next".equals(option))
-			{
-				navigateToSubmission(rundata, "nextSubmissionId");
-			}
-			else if ("prev".equals(option))
-			{
-				navigateToSubmission(rundata, "prevSubmissionId");
-			}
-			else if ("back".equals(option))
+			if ("back".equals(option))
 			{
+				// SAK-29314 - calculate our position relative to the list so we can return to the correct page
+				state.setAttribute(STATE_GOTO_PAGE, calcPageFromSubmission(state));
 				state.setAttribute(STATE_MODE, MODE_INSTRUCTOR_GRADE_ASSIGNMENT);
 			}
 			else if ("backListStudent".equals(option))
 			{
 				state.setAttribute(STATE_MODE, MODE_INSTRUCTOR_VIEW_STUDENTS_ASSIGNMENT);
 			}
+			else if( option.startsWith( FLAG_NEXT ) || option.startsWith( FLAG_PREV ) )
+			{
+				// bjones86 - SAK-29314
+				ParameterParser params = rundata.getParameters();
+				String submissionsOnlySelected = (String) params.getString( PARAMS_VIEW_SUBS_ONLY_CHECKBOX );
+				if( FLAG_ON.equalsIgnoreCase( submissionsOnlySelected ) )
+				{
+					switch( option )
+					{
+						case FLAG_NEXT:
+							navigateToSubmission( rundata, CONTEXT_NEXT_WITH_SUB_ID );
+							break;
+						case FLAG_PREV:
+							navigateToSubmission( rundata, CONTEXT_PREV_WITH_SUB_ID );
+							break;
+						case FLAG_NEXT_UNGRADED:
+							navigateToSubmission( rundata, CONTEXT_NEXT_UNGRADED_WITH_SUB_ID );
+							break;
+						case FLAG_PREV_UNGRADED:
+							navigateToSubmission( rundata, CONTEXT_PREV_UNGRADED_WITH_SUB_ID );
+							break;
+					}
+				}
+				else
+				{
+					switch( option )
+					{
+						case FLAG_NEXT:
+							navigateToSubmission( rundata, "nextSubmissionId" );
+							break;
+						case FLAG_PREV:
+							navigateToSubmission( rundata, "prevSubmissionId" );
+							break;
+						case FLAG_NEXT_UNGRADED:
+							navigateToSubmission( rundata, CONTEXT_NEXT_UNGRADED_SUB_ID );
+							break;
+						case FLAG_PREV_UNGRADED:
+							navigateToSubmission( rundata, CONTEXT_PREV_UNGRADED_SUB_ID );
+							break;
+					}
+				}
+			}
 		}
 
 	} // doPrev_back_next_submission
 
+	/**
+	 * calculate the page of the submission list that the current submission belongs to
+	 * 
+	 * @author plukasew - SAK-29314
+	 * @param state
+	 * @return 
+	 */
+	private Integer calcPageFromSubmission(SessionState state)
+	{
+		int pageSize = Integer.parseInt(state.getAttribute(STATE_PAGESIZE).toString());
+		String submissionId = state.getAttribute(GRADE_SUBMISSION_SUBMISSION_ID).toString();
```

Comment:
I'm noticing we have a few places where we get a submission id / reference from the state, then iterate through all the submissions to find its index in the submission list. If this requires a DB query for each submission, we may want to store the index of the currently viewed submission in the state.